### PR TITLE
Add support for async rules

### PR DESCRIPTION
**Pull request by ramasilveyra**
PR with the intention of solve #2296: add support for async rules.

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@ramasilveyra need add test :+1:

**Comment by ramasilveyra** [comment]
@evilebottnawi Tests added! 😀

**Comment by ramasilveyra** [comment]
@davidtheclark Docs added :) My only concern it's that i'm not sure if I should have called this new doc section as `Asynchronous rules` or `Asynchronous plugins`...

File: docs/developer-guide/plugins.md [hunk 1]
```diff
@@ -36,6 +36,39 @@ In order for your plugin rule to work with the [standard configuration format](.
 
 `ruleFunction` should return a function that is essentially a little [PostCSS plugin](https://github.com/postcss/postcss/blob/master/docs/writing-a-plugin.md): it takes 2 arguments: the PostCSS Root (the parsed AST), and the PostCSS LazyResult. You'll have to [learn about the PostCSS API](https://github.com/postcss/postcss/blob/master/docs/api.md).
 
+### Asynchronous rules
+
+Rules with asynchronous PostCSS plugins are also possible you only need to return a Promise instance.
```
Comment (davidtheclark) [inline docs/developer-guide/plugins.md]: Let's add some punctuation here:

> Rules with asynchronous PostCSS plugins are also possible! All you need to do is return a Promise instance from your plugin function.

**Comment by davidtheclark** [comment]
@ramasilveyra I made the copy edit suggestions directly — realized that I could do that easily with the GitHub UI! Will merge on CI green light. Thanks again.

**Comment by davidtheclark** [comment]
Added to changelog:

> -   Added: support for asynchronous plugin rules ([#2351](https://github.com/stylelint/stylelint/pull/2351)).

**Comment by ramasilveyra** [comment]
💥 Thanks all for your time!

**Pull request merged**